File: 05-binary_search/00153-Find_Minimum_in_Rotated_Sorted_Array/00153-find_minimum_in_rotated_sorted_array.java
// 153: Find Minimum in Rotated Sorted Array
// https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/


class Solution {
    // SOLUTION
    public int findMin(int[] nums) {
        int l = 0;
        int r = nums.length - 1;
        
        while (l<r && nums[l]>nums[r]) {
            int m = l + (r-l) / 2;

            if (nums[m] > nums[r])
                l = m + 1;
            else
                r = m;
        }

        return nums[l];
    }


    public static void main(String[] args) {
        Solution o = new Solution();
        
        // INPUT
        int[] nums = {3,4,5,1,2};

        // OUTPUT
        var result = o.findMin(nums);
        System.out.println(result);
    }
}

File: 05-binary_search/00153-Find_Minimum_in_Rotated_Sorted_Array/00153-find_minimum_in_rotated_sorted_array.py
# 153: Find Minimum in Rotated Sorted Array
# https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/


class Solution:
    # SOLUTION
    def findMin(self, nums: list[int]) -> int:
        l = 0
        r = len(nums) - 1
        
        while (l<r and nums[l]>nums[r]):
            m = l + (r-l) // 2

            if (nums[m] > nums[r]):
                l = m + 1
            else:
                r = m

        return nums[l]



if __name__ == "__main__":
    o = Solution()

    # INPUT
    nums: list[int] = [3,4,5,1,2]

    # OUTPUT
    result = o.findMin(nums)
    print(result)